### ISSUE TITLE
Provide annotated services a way to return a successful response with alternative status code

### ISSUE THREAD

**Issue by trustin** [line/armeria#1081]
For example:

```java
@Get("/hello")
public String greet() {
    return "world"; // A user wants to send a non-200 response, but how?
}
```

Related: https://github.com/line/centraldogma/pull/141#discussion_r173067177

**Comment by hyangtack** [comment]
I think that we need to introduce some class like Spring's [ResponseEntity](https://docs.spring.io/spring/docs/current/javadoc-api/org/springframework/http/ResponseEntity.html) to make the response flexible.

**Comment by hyangtack** [comment]
@trustin suggested a name `HttpResult<?>`. Thank you!

- Returning customized HTTP response.
```
@Get("/api1")
public HttpResult<SomeObject> api1() {
	final SomeObject someObject = produceSomeObject(...);
	return new HttpResultBuilder<SomeObject>().ok()
			                        		  .header("name", "value")
			                        	      .contentType(MediaType.JSON_UTF_8)
			                                  .body(someObject)	// Set an object.
			                                  .build();
}
```

- Returning customized HTTP response after a future has completed.
```
@Get("/api2")
public HttpResult<SomeObject> api2() {
	final CompletableFuture<SomeObject> someObject = produceSomeObjectLater(...);
	return new HttpResultBuilder<SomeObject>().ok()
			                        		  .header("name", "value")
			                        	      .contentType(MediaType.JSON_UTF_8)
			                                  .body(someObject)	// Set a future.
			                                  .build();
}
```

- Returning customized HTTP response with a publisher. How can we support streaming response? Should the type to be a `List<SomeObject>`?
```
@Get("/api3")
public HttpResult<SomeObject> api3() {
	final Publisher<SomeObject> publisher = ...;
	return new HttpResultBuilder<SomeObject>().ok()
			                        		  .header("name", "value")
			                        	      .contentType(MediaType.JSON_UTF_8)
			                                  .body(someObject)	// Set a publisher.
			                                  .build();
}
```

**Comment by hyangtack** [comment]
- For JSON streaming (MIME type `application/json-seq`): https://tools.ietf.org/html/rfc7464
  - https://jira.spring.io/browse/SPR-16742

- For Server Sent Events (MIME type `text/event-stream`): https://developer.mozilla.org/en-US/docs/Web/API/Server-sent_events/Using_server-sent_events
  - https://www.baeldung.com/spring-server-sent-events

**Comment by hyangtack** [comment]
- It'd be better to support [@ResponseStatus](https://github.com/spring-projects/spring-framework/blob/master/spring-web/src/main/java/org/springframework/web/bind/annotation/ResponseStatus.java) as well. 
- We need to change the method signature of [ResponseConverterFunction](https://github.com/line/armeria/blob/master/core/src/main/java/com/linecorp/armeria/server/annotation/ResponseConverterFunction.java) in order to pass `HttpStatus` and `MediaType` into the converter.

**Comment by hyangtack** [comment]
Closed via #1509
